`rustc_interface` cleanups

nnethercote opened the pull request:
Particularly in and around `--cfg` and `--check-cfg` handling.

r? @oli-obk

nnethercote committed:
Remove an unneeded dependency.

nnethercote committed:
Fix a comment.

nnethercote committed:
Streamline `rustc_interface` tests.

In `test_edition_parsing`, change the
`build_session_options_and_crate_config` call to
`build_session_options`, because the config isn't used.

That leaves a single call site for
`build_session_options_and_crate_config`, so just inline and remove it.

nnethercote committed:
Change `CrateConfig` from `FxIndexSet` to `FxHashSet`.

Because its order doesn't matter. This is well demonstrated by
`to_crate_config`, which creates a `CrateConfig` from an `FxHashSet`.

nnethercote committed:
Optimize `parse_cfgspecs`.

In `parse_cfg`, we now construct a `FxHashSet<String>` directly instead of
constructing a `FxHashSet<Symbol>` and then immediately converting it to a
`FxHashSet<String>`(!)

(The type names made this behaviour non-obvious. The next commit will
make the type names clearer.)

nnethercote committed:
Clean up config mess.

`parse_cfgspecs` and `parse_check_cfg` run very early, before the main
interner is running. They each use a short-lived interner and convert
all interned symbols to strings in their output data structures. Once
the main interner starts up, these data structures get converted into
new data structures that are identical except with the strings converted
to symbols.

All is not obvious from the current code, which is a mess, particularly
with inconsistent naming that obscures the parallel string/symbol data
structures. This commit clean things up a lot.

- The existing `CheckCfg` type is generic, allowing both
  `CheckCfg<String>` and `CheckCfg<Symbol>` forms. This is really
  useful, but it defaults to `String`. The commit removes the default so
  we have to use `CheckCfg<String>` and `CheckCfg<Symbol>` explicitly,
  which makes things clearer.

- Introduces `Cfg`, which is generic over `String` and `Symbol`, similar
  to `CheckCfg`.

- Renames some things.
  - `parse_cfgspecs` -> `parse_cfg`
  - `CfgSpecs` -> `Cfg<String>`, plus it's used in more places, rather
    than the underlying `FxHashSet` type.
  - `CrateConfig` -> `Cfg<Symbol>`.
  - `CrateCheckConfig` -> `CheckCfg<Symbol>`

- Adds some comments explaining the string-to-symbol conversions.

- `to_crate_check_config`, which converts `CheckCfg<String>` to
  `CheckCfg<Symbol>`, is inlined and removed and combined with the
  overly-general `CheckCfg::map_data` to produce
  `CheckCfg::<String>::intern`.

- `build_configuration` now does the `Cfg<String>`-to-`Cfg<Symbol>`
  conversion, so callers don't need to, which removes the need for
  `to_crate_config`.

The diff for two of the fields in `Config` is a good example of the
improved clarity:
```
-    pub crate_cfg: FxHashSet<(String, Option<String>)>,
-    pub crate_check_cfg: CheckCfg,
+    pub crate_cfg: Cfg<String>,
+    pub crate_check_cfg: CheckCfg<String>,
```
Compare that with the diff for the corresponding fields in `ParseSess`,
and the relationship to `Config` is much clearer than before:
```
-    pub config: CrateConfig,
-    pub check_config: CrateCheckConfig,
+    pub config: Cfg<Symbol>,
+    pub check_config: CheckCfg<Symbol>,
```

Urgau commented on compiler/rustc_interface/src/interface.rs:
```diff
@@ -117,19 +121,20 @@ pub fn parse_cfgspecs(
-        let exhaustive_names = !specs.is_empty();
-        let exhaustive_values = !specs.is_empty();
-        let mut check_cfg = CheckCfg { exhaustive_names, exhaustive_values, ..CheckCfg::default() };
+        let mut exhaustive_names = !specs.is_empty();
+        let mut exhaustive_values = !specs.is_empty();
+        let mut expecteds = FxHashMap::default();
```
This makes me a bit sad. I found operating on a `CheckCfg` struct to be clearer.
But I suppose it's for the great good.

oli-obk replied:
I think I would prefer keeping around the struct from the start, too. It may be good to add some clearer methods instead (e.g. deduplicating the `old_syntax == None` and immediate overwrite check, or adding methods for inserting new elements to the `expecteds` field).

nnethercote replied:
I'll just remove the last commit, it's not that important.

Urgau commented on compiler/rustc_session/src/config.rs:
```diff
@@ -1385,8 +1383,8 @@ impl<T> Default for CheckCfg<T> {
-impl<T> CheckCfg<T> {
-    fn map_data<O: Eq + Hash>(self, f: impl Fn(T) -> O) -> CheckCfg<O> {
+impl CheckCfg<String> {
+    pub fn intern(self) -> CheckCfg<Symbol> {
```
:+1:

Urgau commented on compiler/rustc_session/src/config.rs:
```diff
@@ -1441,14 +1439,7 @@ impl<'a, T: Eq + Hash + Copy + 'a> Extend<&'a T> for ExpectedValues<T> {
     }
 }
 
-/// Converts the crate `--check-cfg` options from `String` to `Symbol`.
-/// `rustc_interface::interface::Config` accepts this in the compiler configuration,
-/// but the symbol interner is not yet set up then, so we must convert it later.
```
Do you think it is worth keeping this comment? by moving it to the `intern` function maybe.

nnethercote replied:
There are other new comments that communicate the same ideas, so I think that's good enough.

Urgau reviewed (commented):
Thank you. This brings some consistency between the two, which is nice.

nnethercote commented:
Best reviewed one commit at a time.

bjorn3 commented on compiler/rustc_interface/src/lib.rs:
```diff
@@ -24,8 +24,6 @@ mod proc_macro_decls;
 mod queries;
 pub mod util;
 
-pub use callbacks::setup_callbacks;
```
Looks like Geobacter uses this too: https://github.com/geobacter-rs/geobacter/blob/b569ef15dc7d49816e3ae5ea730d09fc94ee7e4b/runtime-core/src/codegen/worker/mod.rs#L86 It's been almost three years since they last updated their rustc fork though it seems.

nnethercote replied:
I will just remove the relevant commit, it's not that important.

Are there any basic principles I can follow to understand what APIs might be used outside of the repository? I've had trouble with things in `rustc_driver_impl` and `rustc_interface`.

bjorn3 replied:
> Are there any basic principles I can follow to understand what APIs might be used outside of the repository?

I did a github search for the function name and `language:rust` and the manually ignored all search results for rustc forks.

nnethercote replied:
Right, so that's a 100% descriptive approach, but it seems like there's nothing prescriptive that can be relied on? "Just don't change anything that anyone is actually using, as determined by GitHub search" is not a good state of affairs. It's very easy to violate unintentionally, and gives no guidance on how to change things that are relied on.

oli-obk replied:
Usually we can remove everything that is not used in rustc and doesn't have a comment explaining which external tool needs it.

nnethercote replied:
My experience is that I do that, and then someone (often @bjorn3) complains, and so then I put it back and add that comment :/

nnethercote replied:
#116827 is the most recent example.

nnethercote committed:
Change `Cfg<T>` to an `FxIndexSet`.

Despite what I claimed in an earlier commit, the ordering does matter to
some degree. Using `FxIndexSet` prevents changes to the error message
order in `tests/ui/check-cfg/mix.rs`.

nnethercote commented:
I removed the two controversial commits. I also added a new commit changing `Cfg<T>` back to an `FxIndexSet`, because I got an error message order difference after updating, which suggests that `FxHashSet` isn't good enough. I think it should be ready for re-review.

oli-obk commented:
@bors r+

bors commented:
:pushpin: Commit 5438004766fc0f9b123608425b8ec7111f536640 has been approved by `oli-obk`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=oli-obk 5438004766fc0f9b123608425b8ec7111f536640 -->
<!-- homu: {"type":"Approved","sha":"5438004766fc0f9b123608425b8ec7111f536640","approver":"oli-obk","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->
